src/rgw/driver/rados/role.cc

```diff
@@ -0,0 +1,720 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright contributors to the Ceph project
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#include "role.h"
+
+#include <optional>
+#include <variant>
+
+#include "common/errno.h"
+#include "rgw_common.h"
+#include "rgw_metadata.h"
+#include "rgw_metadata_lister.h"
+#include "rgw_role.h"
+#include "rgw_string.h"
+#include "rgw_tools.h"
+#include "rgw_zone.h"
+#include "svc_mdlog.h"
+
+#include "account.h"
+#include "roles.h"
+
+namespace rgwrados::role {
+
+// RGWRoleInfo is stored in rados objects named "roles.{id}",
+// where ids are assumed to be globally unique
+static const std::string oid_prefix = "roles.";
+// read_by_name() is enabled by rados objects
+// "{tenant}role_names.{name}" for tenant roles, or
+// "{account}role_names.{name}" for account roles
+constexpr std::string_view name_oid_prefix = "role_names.";
+// list() by path/prefix is enabled by rados objects
+// "{tenant}role_paths.{path}roles.{id}" for tenant roles.
+// see rgwrados::roles::list() for account roles
+constexpr std::string_view path_oid_prefix = "role_paths.";
+
+
+static rgw_raw_obj get_id_obj(const RGWZoneParams& zone,
+                              std::string_view id)
+{
+  return {zone.roles_pool, string_cat_reserve(oid_prefix, id)};
+}
+
+static int read_info(const DoutPrefixProvider* dpp, optional_yield y,
+                     RGWSI_SysObj& sysobj, const rgw_raw_obj& obj,
+                     RGWRoleInfo& info, ceph::real_time* pmtime,
+                     RGWObjVersionTracker* pobjv,
+                     rgw_cache_entry_info* pcache_info)
+{
+  bufferlist bl;
+  std::map<std::string, bufferlist> attrs;
+  // "tagging" doesn't start with RGW_ATTR_PREFIX, don't filter it out
+  constexpr bool raw_attrs = true;
+  int r = rgw_get_system_obj(&sysobj, obj.pool, obj.oid, bl, pobjv,
+                             pmtime, y, dpp, &attrs, pcache_info,
+                             boost::none, raw_attrs);
```

Comment:
> still having trouble with the sts test case test_assume_role_with_web_identity_role_resource_tag, which seems to fail because it doesn't find the role tags

turns out this is because i didn't ask for `raw_attrs`, so the cache was filtering out the "tagging" attr